classes/requests/order/session/class-ledyer-create-hpp.php

```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * Create HPP
+ *
+ * @package Ledyer\Requests\Order\Session
+ */
+
+namespace Ledyer\Requests\Order\Session;
+
+use Ledyer\Requests\Order\Request_Order;
+
+defined( 'ABSPATH' ) || exit();
+
+/**
+ * Class Create_HPP
+ *
+ * @package Ledyer\Requests\Order\Session
+ */
+class Create_HPP extends Request_Order {
+	/**
+	 * Adds a confirmation URL to the cart data.
+	 *
+	 * @param array $data The WooCommerce cart data.
+	 * @param int   $order_id The WooCommerce order id.
+	 * @return array The modified cart data.
+	 */
+	public function set_confirmation_url( $data, $order_id ) {
+		$merchant_urls = ledyer()->merchant_urls->get_urls( $order_id );
+		if ( $merchant_urls ) {
+			$data['settings']['urls']['confirmation'] = $merchant_urls['confirmation'];
+		}
+		return $data;
+	}
+
+	/**
+	 * Creates the HPP URL.
+	 *
+	 * @param string $session_id The Ledyer Checkout session to use for the HPP request.
+	 * @return string The HPP URL.
+	 */
+	public function create_hpp_url( $session_id ) {
+		// Todo: remove hardcoding here.
+		return 'https://pos.sandbox.ledyer.com/?sessionId=' . $session_id;
+	}
+
+	/**
+	 *
+	 * Request method
+	 */
+	protected function set_url() {
+		$this->url = sprintf( 'v1/sessions/%s', $this->arguments['orderId'] );
```

Comment:
Only a suggestion, but this could also be written as:

```php
$this->url = "v1/sessions/{$this->arguments['orderId']}";
```

Reply:
This is copied from the other classes that inherit from Request_Order and I'm honestly too bad at OOP to know why I need it, I just know that I do not seem to be allowed to skip it in the class :P I assume it's due to some setting in Request_Order?

Reply:
No problem! I can explain it. The class you created `Create_HPP` inherits (extends) the `Request_Order` class which has an [abstract function](https://github.com/krokedil/ledyer-checkout-for-woocommerce/blob/develop-hpp-redirect/classes/requests/order/class-ledyer-request-order.php#L45). This says, `Request_Order` expects a class that extends it to have a function called `set_url`, and "I don't care how it is implemented, I just expect it to be defined". This applies to all abstract functions, and it is useful when you want classes to have a similar API, but with perhaps different implementation details. You can also have concrete functions such as the [`set_request_url`](https://github.com/krokedil/ledyer-checkout-for-woocommerce/blob/develop-hpp-redirect/classes/requests/order/class-ledyer-request-order.php#L22-L41) that is already defined for you, and available in all subclasses. It is considered "concrete" since it is not "abstract".

Furthermore, since it has abstract functions, the class itself must also be an `abstract class`. This means you cannot instantiate it. That is, you cannot do `new Request_Order()` which forces you extend it to create a concrete class such as `Create_HPP`.

---
In this case, it seems like the `set_url` exists for overriding the `$this->request_url` property. Probably, this is how it is supposed to look:

```php
protected function set_url() {
    $this->url = "{$this->request_url}v1/sessions/{$this->arguments['orderId']}";
}
```

Reply:
Thank you for a great explanation! :) Do you think that I should adjust the content of the method?